x-element.js

```diff
@@ -1,15 +1,721 @@
+      if (attributes.has(attribute)) {
+        throw new Error(`${path}.${key} causes a duplicated attribute "${attribute}".`);
+      }
+      attributes.add(attribute);
+      const { dependencies } = property;
+      if (dependencies) {
+        dependencyMap.set(property, dependencies.map(dependencyKey => properties[dependencyKey]));
+        for (const [index, dependencyKey] of Object.entries(dependencies)) {
+          if (XElement.__typeIsWrong(Object, properties[dependencyKey])) {
+            throw new Error(`${path}.${key}.dependencies[${index}] has an unexpected item ("${dependencyKey}" has not been declared).`);
+          }
+        }
+      }
+    }
+    for (const [key, property] of entries) {
+      if (XElement.__propertyIsCyclic(property, dependencyMap)) {
+        throw new Error(`${path}.${key}.dependencies are cyclic.`);
+      }
+    }
+  }
+
+  static __validateProperty(constructor, key, property) {
+    const path = `${constructor.name}.properties.${key}`;
+    if (X_ELEMENT_INTERFACE.has(key)) {
+      throw new Error(`Unexpected key "${path}" shadows XElement.prototype interface (${[...X_ELEMENT_INTERFACE].join(', ')}).`);
+    }
+    if (INHERITED_INTERFACE.has(key)) {
+      const error = new Error(`Unexpected key "${path}" shadows inherited property name, behavior not guaranteed.`);
+      console.warn(error); // eslint-disable-line no-console
+    }
+    for (const propertyKey of Object.keys(property)) {
+      if (PROPERTY_KEYS.has(propertyKey) === false) {
+        throw new Error(`Unexpected key "${path}.${propertyKey}".`);
+      }
+    }
+    const { attribute, resolver, dependencies, reflect, internal, readOnly, value } = property;
+    for (const subKey of ['type', 'resolver', 'observer']) {
+      if (Reflect.has(property, subKey) && XElement.__typeIsWrong(Function, property[subKey])) {
+        const typeName = XElement.__getTypeName(property[subKey]);
+        throw new Error(`Unexpected value for "${path}.${subKey}" (expected Function, got ${typeName}).`);
+      }
+    }
+    for (const subKey of ['reflect', 'internal', 'readOnly']) {
+      if (Reflect.has(property, subKey) && XElement.__typeIsWrong(Boolean, property[subKey])) {
+        const typeName = XElement.__getTypeName(property[subKey]);
+        throw new Error(`Unexpected value for "${path}.${subKey}" (expected Boolean, got ${typeName}).`);
+      }
+    }
+    if (Reflect.has(property, 'attribute') && XElement.__typeIsWrong(String, attribute)) {
+      const typeName = XElement.__getTypeName(attribute);
+      throw new Error(`Unexpected value for "${path}.attribute" (expected String, got ${typeName}).`);
+    }
+    if (Reflect.has(property, 'attribute') && attribute === '') {
+      throw new Error(`Unexpected value for "${path}.attribute" (expected non-empty String).`);
+    }
+    if (
+      Reflect.has(property, 'value') &&
+      value !== undefined &&
+      value !== null &&
+      XElement.__typeIsWrong(Boolean, value) &&
```

Comment:
This check threw me for a loop; I was trying to assign `new URL('iframe/index.html', import.meta.url)` as my value without using a function, since I don't mind if instances share the URL object. The error message was confusing because I first thought it was referring to the URL type I had declared as the property type.